### PR TITLE
dedupe assignment module

### PR DESCRIPTION
**Pull request by jcwuzoegiver**
1. Consolidated 2 generateAssignment() methods into one.
2. Eliminated redundant code and unused
mergePersonalizationResponseWithSegmentation() method
3. Updated getBucketList() method so it will call AssignmentDecorator to
retrieve recommended bucketList.

@mans4singh @bargenilesh

### CODE REVIEW

File: modules/assignment/src/test/java/com/intuit/wasabi/assignment/impl/DefaultAssignmentDecoratorTest.java [hunk 1]
```diff
@@ -37,11 +40,12 @@
         assertThat(defaultAssignmentDecorator, is(not(nullValue())));
     }
 
-    @Test public void materializeUriTest() throws UnsupportedEncodingException {
+    @Test public void getBucketListTest() throws UnsupportedEncodingException {
```
Comment (mans4singh) [inline modules/assignment/src/test/java/com/intuit/wasabi/assignment/impl/DefaultAssignmentDecoratorTest.java]: Do need test for positive case (when bucket list returned is not null) ?

File: modules/assignment/src/main/java/com/intuit/wasabi/assignment/AssignmentDecorator.java [hunk 2]
```diff
@@ -15,23 +15,38 @@
...
-import java.net.URI;
+import com.intuit.wasabi.assignmentobjects.SegmentationProfile;
+import com.intuit.wasabi.assignmentobjects.User;
+import com.intuit.wasabi.exceptions.BucketDistributionNotFetchableException;
+import com.intuit.wasabi.experimentobjects.BucketList;
+import com.intuit.wasabi.experimentobjects.Experiment;
 
 /**
- * Decorator for creating uri from experiment
+ * Interface that provides method(s) to modify/enhance the default behavior of retrieving bucket list when
```
Comment (mans4singh) [inline modules/assignment/src/main/java/com/intuit/wasabi/assignment/AssignmentDecorator.java]: Should we name the interface BucketSelector - since it is just getting a bucket list ?

Comment (jcwuzoegiver) [inline modules/assignment/src/main/java/com/intuit/wasabi/assignment/AssignmentDecorator.java]: @mans4singh probably not. Today this interface AssignmentDecorator can "decorate" bucket list. Tomorrow, we can add another method to "decorate" the assignment. Let's keep the same name for now. agreed?

File: modules/assignment/src/main/java/com/intuit/wasabi/assignment/impl/AssignmentsImpl.java [hunk 3]
```diff
@@ -963,24 +939,46 @@ private Assignment generateAssignment(Experiment experiment, User.ID userID, Con
...
      */
-    protected BucketList getBucketList(Experiment experiment, Boolean skipBucketRetrieval) {
-        BucketList buckets = null;
-        if (!skipBucketRetrieval) {
-            buckets = repository.getBuckets(experiment.getID());
+    protected BucketList getBucketList(Experiment experiment, User.ID userID, SegmentationProfile segmentationProfile, Boolean skipBucketRetrieval) {
+        
+        BucketList bucketList = null;
+        Boolean isPersonalizationEnabled = experiment.getIsPersonalizationEnabled();
+        if (isPersonalizationEnabled && !Objects.isNull(assignmentDecorator)) {
```
Comment (bargenilesh) [inline modules/assignment/src/main/java/com/intuit/wasabi/assignment/impl/AssignmentsImpl.java]: Very minor: No need of this condition "!Objects.isNull(assignmentDecorator)" as assignmentDecorator should not be NULL in any case...

File: modules/assignment/src/main/java/com/intuit/wasabi/assignment/impl/AssignmentsImpl.java [hunk 4]
```diff
@@ -963,24 +939,46 @@ private Assignment generateAssignment(Experiment experiment, User.ID userID, Con
...
+            try {
+                bucketList = assignmentDecorator.getBucketList(experiment, userID, segmentationProfile);
+            } catch (BucketDistributionNotFetchableException | IOException e) {
+                LOGGER.error(new StringBuilder("Error obtaining the Bucket List from ").append(assignmentDecorator.getClass().getSimpleName())
```
Comment (bargenilesh) [inline modules/assignment/src/main/java/com/intuit/wasabi/assignment/impl/AssignmentsImpl.java]: Can we add 'e' exception variable as second parameter to the LOGGER.error() method, that will print complete stack trace which will be helpful for investigation...

File: modules/assignment/src/test/java/com/intuit/wasabi/assignment/impl/DefaultAssignmentDecoratorTest.java (hunk 1 above)
Comment (jcwuzoegiver) [inline modules/assignment/src/test/java/com/intuit/wasabi/assignment/impl/DefaultAssignmentDecoratorTest.java]: @mans4singh DefaultAssignmentDecorator's getBucketList() always return null;

File: modules/assignment/src/main/java/com/intuit/wasabi/assignment/impl/AssignmentsImpl.java (hunk 4 above)
Comment (jcwuzoegiver) [inline modules/assignment/src/main/java/com/intuit/wasabi/assignment/impl/AssignmentsImpl.java]: @bargenilesh going to eliminate IOException altogether from AssignmentDecorator.getBucketList() interface. This method should not throw IOException. It should be abstract out from the actual implementation of this method by sub class.

**Comment by mans4singh** [comment]
:shipit:
